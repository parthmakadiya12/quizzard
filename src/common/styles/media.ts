const size: any = {
  mobileS: 320,
  mobileM: 375,
  mobileL: 425,
  tablet: 768,
  laptop: 1024,
  laptopL: 1440,
  desktop: 2560,
};
const media: any = Object.keys(size).reduce((acc: any, cur: any) => {
  acc[cur] = `(min-width: ${size[cur]}px)`;
  return acc;
}, {});
export default media;
