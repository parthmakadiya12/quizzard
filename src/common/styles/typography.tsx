import styled from "styled-components";

export const H1 = styled.h1`
  font-weight: 300;
  font-size: 2rem;
`;

export const H2 = styled.h2`
  font-weight: 300;
  font-size: 1.5rem;
`;

export const H3 = styled.h3`
  font-weight: 300;
  font-size: 1.2rem;
`;

export const H4 = styled.h4`
  font-weight: 300;
  font-size: 1rem;
`;

export const P = styled.p`
  font-weight: 300;
  font-size: 0.8rem;
  margin: 0;
  padding: 0;
`;
