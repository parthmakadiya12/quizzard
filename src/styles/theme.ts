import { createGlobalStyle } from "styled-components";

export const darkTheme = {
  color: {
    body: "#262c4a",
    cardBackGround: "#262c4a",
    fontPrimary: "#fdfefe", // white shade
    fontSecondry: "#cec3c3", // dirty white
    headerColor: "#8d95bb", // light purple
    checkBox: "#117fec", // blue
    title: "#8c94bd", // wild blue
    button: "#127eeb", //light blue
    buttonSecondry: "#4098f0", //lighter blue
    buttonFontColor: "white",
    inputColor: "#4a5171",
    inputFontColor: "white",
  },
  type: "dark",
};

export const lightTheme = {
  color: {
    body: "#ffda62",
    cardBackGround: "#ffffff",
    fontPrimary: "#1d345b",
    fontSecondry: "#1A1A1A",
    headerColor: "black",
    checkBox: "#ff5261",
    title: "#1f345a",
    button: "#6287FF",
    buttonSecondry: "#2F60FF",
    buttonFontColor: "white",
    inputColor: "white",
    inputFontColor: "#1d345b",
  },
  type: "light",
};

export const GlobalStyles = createGlobalStyle<{ theme: any }>`
    body{
        background-color: ${({ theme }) => theme.color.body};
        color:${({ theme }) => theme.color.fontPrimary}
    }
`;
