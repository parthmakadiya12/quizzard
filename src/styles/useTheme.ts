import { useState } from 'react'

const useTheme=():[string,()=>void]=> {
    const [theme,setTheme]=useState("dark");

    const toggleTheme=():void=>{
        theme==="dark"? setTheme("light"):setTheme("dark");
    }
    
    return [theme,toggleTheme];
}

export default useTheme;