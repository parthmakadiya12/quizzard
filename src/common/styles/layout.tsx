import styled, { css } from "styled-components";
import media from "./media";

export const Row = styled.div<any>`
  display: flex;
  flex-direction: row;
  ${(props) =>
    props.justifyContent &&
    css`
      justify-content: ${props.justifyContent};
    `}
  ${(props) =>
    props.alignItems &&
    css`
      align-items: ${props.alignItems};
    `}
`;

export const Col = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Wrapper = styled(Col)`
  background-color: ${({ theme }) => theme.color.cardBackGround};
  border: 3px solid #21486b;
  border-radius: 0.8rem;
  padding: 2rem;
  margin: 0.5rem;

  margin: auto;
  ${({ theme }) =>
    theme.type === "light" &&
    `
      border:none;
      box-shadow: rgba(17, 17, 26, 0.05) 0px 4px 16px, rgba(17, 17, 26, 0.05) 0px 8px 32px;
  `}
  @media ${media.tablet} {
    width: 60%;
  }
`;

export const BigQuestionWrapper = styled(Wrapper)`
  margin: 0 10px;
  @media ${media.tablet} {
    max-width: 600px;
    margin: auto;
  }
`;

export const CardTitle = styled.span`
  color: ${({ theme }) => theme.color.headerColor};
  width: 100%;
  font-size: 1.4rem;
  font-weight: 500;
  margin-bottom: 1rem;
  padding-bottom: 1rem;
  background-image: ${({ theme }) =>
    `linear-gradient(to right, ${theme.color.headerColor} 33%, rgba(255,255,255,0) 0%)`};
  background-position: bottom;
  background-size: 0.5rem 0.2rem;
  background-repeat: repeat-x;
`;

export const SubTitle = styled.span`
  font-size: 1rem;
  text-transform: capitalize;
`;

export const SmallText = styled.span`
  font-size: 0.8rem;
  margin-bottom: 0.5rem;
  color: ${({ theme }) => theme.color.fontSecondry};
`;
