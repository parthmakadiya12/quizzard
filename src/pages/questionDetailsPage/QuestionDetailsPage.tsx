import { FC, useState, useEffect } from "react";
import styled from "styled-components";
import { useParams } from "react-router-dom";

import { CardTitle, Wrapper, Row } from "../../common/styles/layout";
import { P } from "../../common/styles/typography";
import { Radio } from "../../common/Radio/Radio";
import { Button } from "../../common";
import { QuestionType } from "../../common/types/QuestionType";
import media from "../../common/styles/media";

export interface PropTypes {
  question?: QuestionType;
  getQuestionData: (id: any) => void;
  voteOnQuestion: (name: string, url: string) => void;
}
const QuestionDetailsPage: FC<PropTypes> = ({
  question,
  voteOnQuestion,
  getQuestionData,
}) => {
  const { questionId } = useParams<{ questionId: string }>();
  const [name, setName] = useState<string>("");
  const [choiceUrl, setChoiceUrl] = useState<string>("");
  const [quesUrl, setQuesUrl] = useState<string>("");

  const clickHandler = (name: string, url: string, queUrl: string) => {
    setName(name);
    setChoiceUrl(url);
    setQuesUrl(queUrl);
  };

  const handleSave = (quesUrl: string, choiceUrl: string) => {
    if (!!quesUrl && !!choiceUrl) {
      voteOnQuestion(quesUrl, choiceUrl);
    }
  };

  useEffect(() => {
    if (!question) {
      getQuestionData(questionId);
    }
  }, []);

  return (
    <WrapperStyled id="questionWrapper">
      <CardTitle>{question && question.question}</CardTitle>
      {question &&
        question.choices.map((option: any) => (
          <Row key={`${option.url}_row`} alignItems={"center"}>
            <Radio
              name={option.choice}
              choiceUrl={option.url}
              queUrl={question.url}
              key={`${option.url}_radio`}
              clickHandler={(name: string, choiceUrl: string, queUrl: string) =>
                clickHandler(name, choiceUrl, queUrl)
              }
              checked={false}
              selected={name}
            />
            <P key={`${option.url}_text`}>{option.percentage}%</P>
          </Row>
        ))}
      <Button clickHandler={() => handleSave(quesUrl, choiceUrl)}>Save</Button>
    </WrapperStyled>
  );
};

const WrapperStyled = styled(Wrapper)`
  & > button {
    margin-top: 1rem;
  }
  @media ${media.tablet} {
    max-width: 600px;
  }
`;

export default QuestionDetailsPage;
