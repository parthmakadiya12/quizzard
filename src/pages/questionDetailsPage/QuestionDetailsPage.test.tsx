import renderer from "react-test-renderer";
import QuestionDetailsPage, { PropTypes } from "./QuestionDetailsPage";
import { ThemeProvider } from "styled-components";
import { darkTheme } from "../../styles/theme";
import "jest-styled-components";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    questionId: "12",
  }),
  useRouteMatch: () => ({ url: "/questions/12/" }),
}));

describe("<QuestionsDetailsPage>", () => {
  let props: PropTypes;
  beforeEach(() => {
    props = {
      question: {
        choices: [
          {
            choice: "Swift",
            percentage: 53.33,
            url: "/questions/1/choices/1",
            votes: 2048,
          },
          {
            choice: "Python",
            percentage: 26.67,
            url: "/questions/1/choices/2",
            votes: 1024,
          },
          {
            choice: "Objective-C",
            percentage: 13.33,
            url: "/questions/1/choices/3",
            votes: 512,
          },
          {
            choice: "Ruby",
            percentage: 6.67,
            url: "/questions/1/choices/4",
            votes: 256,
          },
        ],
        published_at: "2014-11-11T08:40:51.620Z",
        question: "Favourite programming language?",
        url: "/questions/1",
      },
      getQuestionData: jest.fn(),
      voteOnQuestion: jest.fn(),
    };
  });
  it("should render correctly", () => {
    const wrapper = mount(
      <ThemeProvider theme={darkTheme}>
        <QuestionDetailsPage {...props} />
      </ThemeProvider>
    ).debug();
    expect(wrapper).toMatchSnapshot();
  });

  it("should not call voteOnQuestion if none of the radio is selected", () => {
    const wrapper = mount(
      <ThemeProvider theme={darkTheme}>
        <QuestionDetailsPage {...props} />
      </ThemeProvider>
    );
    wrapper.find("button").simulate("click");
    expect(props.voteOnQuestion).not.toHaveBeenCalled();
  });

  it("should call voteOnQuestion by clicking vote button", () => {
    const wrapper = mount(
      <ThemeProvider theme={darkTheme}>
        <QuestionDetailsPage {...props} />
      </ThemeProvider>
    );
    wrapper.find("span").at(1).simulate("click");
    wrapper.find("button").simulate("click");
    expect(props.voteOnQuestion).toHaveBeenCalledWith(
      "/questions/1/choices/1",
      "/questions/1"
    );
  });

  it("should call getQuestionData if user visits child page directly", () => {
    const newProps = { ...props, question: undefined };
    const wrapper = mount(
      <ThemeProvider theme={darkTheme}>
        <QuestionDetailsPage {...newProps} />
      </ThemeProvider>
    );

    expect(props.getQuestionData).toHaveBeenCalled();
  });
});
