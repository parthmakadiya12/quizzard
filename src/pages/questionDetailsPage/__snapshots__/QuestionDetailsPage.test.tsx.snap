// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuestionsDetailsPage> should render correctly 1`] = `
"<Component theme={{...}}>
  <QuestionDetailsPage question={{...}} getQuestionData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} voteOnQuestion={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
    <Styled(Styled(styled.div)) id=\\"questionWrapper\\">
      <div id=\\"questionWrapper\\" className=\\"sc-gtsrHT sc-dlnjwi sc-bqGGPW cPlHGU eogWRV fbxJK\\">
        <styled.span>
          <span className=\\"sc-hKFxyN eTEXUW\\">
            Favourite programming language?
          </span>
        </styled.span>
        <styled.div alignItems=\\"center\\">
          <div className=\\"sc-bdnxRM oNYUZ\\">
            <Radio name=\\"Swift\\" choiceUrl=\\"/questions/1/choices/1\\" queUrl=\\"/questions/1\\" clickHandler={[Function: clickHandler]} checked={false} selected=\\"\\">
              <div>
                <styled.div>
                  <div className=\\"sc-pNWdM zdBa-d\\">
                    <styled.input value=\\"Swift\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false}>
                      <input value=\\"Swift\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false} className=\\"sc-jrsJWt YpWzs\\" />
                    </styled.input>
                    <styled.span onClick={[Function: onClick]} Checked={false}>
                      <span onClick={[Function: onClick]} className=\\"sc-kEqXSa bmqgSv\\">
                        Swift
                      </span>
                    </styled.span>
                  </div>
                </styled.div>
              </div>
            </Radio>
            <styled.p>
              <p className=\\"sc-fujyAs dJRJQl\\">
                53.33
                %
              </p>
            </styled.p>
          </div>
        </styled.div>
        <styled.div alignItems=\\"center\\">
          <div className=\\"sc-bdnxRM oNYUZ\\">
            <Radio name=\\"Python\\" choiceUrl=\\"/questions/1/choices/2\\" queUrl=\\"/questions/1\\" clickHandler={[Function: clickHandler]} checked={false} selected=\\"\\">
              <div>
                <styled.div>
                  <div className=\\"sc-pNWdM zdBa-d\\">
                    <styled.input value=\\"Python\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false}>
                      <input value=\\"Python\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false} className=\\"sc-jrsJWt YpWzs\\" />
                    </styled.input>
                    <styled.span onClick={[Function: onClick]} Checked={false}>
                      <span onClick={[Function: onClick]} className=\\"sc-kEqXSa bmqgSv\\">
                        Python
                      </span>
                    </styled.span>
                  </div>
                </styled.div>
              </div>
            </Radio>
            <styled.p>
              <p className=\\"sc-fujyAs dJRJQl\\">
                26.67
                %
              </p>
            </styled.p>
          </div>
        </styled.div>
        <styled.div alignItems=\\"center\\">
          <div className=\\"sc-bdnxRM oNYUZ\\">
            <Radio name=\\"Objective-C\\" choiceUrl=\\"/questions/1/choices/3\\" queUrl=\\"/questions/1\\" clickHandler={[Function: clickHandler]} checked={false} selected=\\"\\">
              <div>
                <styled.div>
                  <div className=\\"sc-pNWdM zdBa-d\\">
                    <styled.input value=\\"Objective-C\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false}>
                      <input value=\\"Objective-C\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false} className=\\"sc-jrsJWt YpWzs\\" />
                    </styled.input>
                    <styled.span onClick={[Function: onClick]} Checked={false}>
                      <span onClick={[Function: onClick]} className=\\"sc-kEqXSa bmqgSv\\">
                        Objective-C
                      </span>
                    </styled.span>
                  </div>
                </styled.div>
              </div>
            </Radio>
            <styled.p>
              <p className=\\"sc-fujyAs dJRJQl\\">
                13.33
                %
              </p>
            </styled.p>
          </div>
        </styled.div>
        <styled.div alignItems=\\"center\\">
          <div className=\\"sc-bdnxRM oNYUZ\\">
            <Radio name=\\"Ruby\\" choiceUrl=\\"/questions/1/choices/4\\" queUrl=\\"/questions/1\\" clickHandler={[Function: clickHandler]} checked={false} selected=\\"\\">
              <div>
                <styled.div>
                  <div className=\\"sc-pNWdM zdBa-d\\">
                    <styled.input value=\\"Ruby\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false}>
                      <input value=\\"Ruby\\" type=\\"radio\\" name=\\"radio\\" defaultChecked={false} className=\\"sc-jrsJWt YpWzs\\" />
                    </styled.input>
                    <styled.span onClick={[Function: onClick]} Checked={false}>
                      <span onClick={[Function: onClick]} className=\\"sc-kEqXSa bmqgSv\\">
                        Ruby
                      </span>
                    </styled.span>
                  </div>
                </styled.div>
              </div>
            </Radio>
            <styled.p>
              <p className=\\"sc-fujyAs dJRJQl\\">
                6.67
                %
              </p>
            </styled.p>
          </div>
        </styled.div>
        <Button clickHandler={[Function: clickHandler]}>
          <styled.button onClick={[Function: clickHandler]}>
            <button onClick={[Function: clickHandler]} className=\\"sc-crzoAE gyLjuP\\">
              Save
            </button>
          </styled.button>
        </Button>
      </div>
    </Styled(Styled(styled.div))>
  </QuestionDetailsPage>
</Component>"
`;
